### PR TITLE
feat(network): add single node networking capabilities

### PR DESCRIPTION
**Pull request by lucido-simon**
This PR defines a new crate, to be used by the agent, when creating a new node or a new instance.

It allows :
- [x] Creating a bridge to be used by virtual interfaces later on
- [x] Setting up network configuration, and iptables rules
- [x] Creating a network namespace to be used by instances
- [x] Creating virtual interfaces for this namespace
- [x] Setting up NAT/PAT rules for this interface
- [x] Cleaning up both nodes & instances 

The documentation has also been updated.

### CODE REVIEW

File: network/src/utils/mod.rs [hunk 1]
```diff
@@ -15,6 +15,29 @@ pub(crate) fn run_command(cmd: &str, args: &[&str]) -> Result<String, KudoNetwor
     wrap_command_output(output, cmd, args)
 }
 
+pub(crate) fn get_default_interface_name() -> Result<String, KudoNetworkError> {
```
Comment (sameo) [inline network/src/utils/mod.rs:18]: getters in rust usually don't have the `get_`prefix.

File: network/src/utils/mod.rs [hunk 2]
```diff
@@ -15,6 +15,29 @@ pub(crate) fn run_command(cmd: &str, args: &[&str]) -> Result<String, KudoNetwor
...
+        Err(e) => Err(KudoNetworkError::DefaultNetworkInterfaceError(format!(
+            "Could not find default interface : {}",
+            e
+        ))),
+    }
+}
+
+pub fn get_namespace_name(instance_id: String) -> String {
+    // Linux network interface names are size limited!
+    format!("kns{}", &instance_id[..min(12, instance_id.len())])
```
Comment (sameo) [inline network/src/utils/mod.rs:30]: 12 or the interface name length should be defined as a constant in your crate.

File: network/src/utils/mod.rs [hunk 3]
```diff
@@ -15,6 +15,29 @@ pub(crate) fn run_command(cmd: &str, args: &[&str]) -> Result<String, KudoNetwor
...
 
+pub(crate) fn get_default_interface_name() -> Result<String, KudoNetworkError> {
+    match default_net::get_default_interface() {
+        Ok(default_interface) => Ok(default_interface.name),
+        Err(e) => Err(KudoNetworkError::DefaultNetworkInterfaceError(format!(
+            "Could not find default interface : {}",
+            e
+        ))),
+    }
+}
```
Comment (sameo) [inline network/src/utils/mod.rs:26]: You should use a `map().map_err()`combination for this.

**Comment by sameo** [comment]
`- delete api.proto, since we're not going the gRPC route anymore` I understand you're going the rust crate way, that's good.
Please use a specific commit that would update the documentation and remove the grpc protos. Dont mix the proto removal with code that adds the new API.

**Review by sameo — COMMENTED** [review]
The PR looks good now, thanks. But your 4th commit containing all the fixes should not be there. All fixes in that commit should be applied to their relevant commts. Use interactive rebase (see `git rebase -i`) for that purpose.

File: network/src/utils/mod.rs (hunk 3 above)
Comment (sameo) [reply network/src/utils/mod.rs:28]: You have not addressed that one yet?

Comment (RedbeanGit) [reply network/src/utils/mod.rs]: My bad, it's corrected now

**Review by sameo — APPROVED** [review]

**Pull request merged**